File: allcode/300-399/331isValidSerialization.py
# 序列化二叉树的一种方法是使用前序遍历。当我们遇到一个非空节点时，我们可以记录下这个节点的值。如果它是一个空节点，我们可以使用一个标记值记录，例如 #。
#
#      _9_
#     /   \
#    3     2
#   / \   / \
#  4   1  #  6
# / \ / \   / \
# # # # #   # #
# 例如，上面的二叉树可以被序列化为字符串 "9,3,4,#,#,1,#,#,2,#,6,#,#"，其中 # 代表一个空节点。
#
# 给定一串以逗号分隔的序列，验证它是否是正确的二叉树的前序序列化。编写一个在不重构树的条件下的可行算法。
#
# 每个以逗号分隔的字符或为一个整数或为一个表示 null 指针的 '#' 。
#
# 你可以认为输入格式总是有效的，例如它永远不会包含两个连续的逗号，比如"1,,3" 。
#
# 示例 1:
#
# 输入: "9,3,4,#,#,1,#,#,2,#,6,#,#"
# 输出: true
# 示例2:
#
# 输入: "1,#"
# 输出: false
# 示例 3:
#
# 输入: "9,#,#,1"
# 输出: false

class Solution:
    def isValidSerialization(self, preorder: str) -> bool:
        stack =[]
        preorderList = preorder.split(',')
        for i, val in enumerate(preorderList):
            if val != '#':
                stack.append(val)
            else:
                if 0 == len(stack):
                    if i < len(preorderList) - 1:
                        return False
                    else:
                        return True
                stack.pop()
        return False




so = Solution()
print('res =', so.isValidSerialization("9,3,4,#,#,1,#,#,2,#,6,#,#"))
print('res =', so.isValidSerialization("1,#"))
print('res =', so.isValidSerialization("9,#,#,1"))

File: c++/src/1-999/331isValidSerialization.cpp
#include "lc_pub.h"


// 用了两种lambda递归函数的写法，一种带this，不能引用成员变量，另一种不带this，能引用成员变量

class Solution {
public:
bool isValidSerialization(string preorder) {
    if(preorder=="#") return true;
    stack<int> stk;
    int n = preorder.size();
    for (int i=0;i<n;i++) {
        if(preorder[i]==',') continue;
        if(preorder[i]=='#') {
            if (stk.size()==0) return false;
            while(stk.size()) {
                if(stk.top()==2) stk.pop();
                else {stk.top()=2;break;}
            }
            if (stk.size()==0) return i==n-1;
        }
        else {
            if (i==0||preorder[i-1]==',') stk.push(1);
            else continue;
        }
    }
    return stk.size()==0;
}
};

int main()
{
    std::cout<<"test let us start! %s" << __cplusplus <<std::endl;

    Solution so;
    auto v = so.isValidSerialization("9,3,4,#,#,1,#,#,2,#,6,#,#");
    cout << v << endl;
    return 0;
}
